### PR TITLE
IMPRO_504 Downscaled precision and added unit tests

### PR DESCRIPTION
**Pull request by MarkWorsfold**
Issue: #435 

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/wind_downscaling/test_roughnesscorrection.py [hunk 1]
```diff
@@ -729,6 +737,17 @@ def test_section2c(self):
...
 
+    def test_section2d(self):
+        """ Test whether output is float32. """
+        hlvs = 10
+        uin = np.ones(hlvs)*20
+        heights = ((np.arange(hlvs)+1)**2.)*12.
+        multip_hc_rc = TestMultiPoint(3)
+        land_hc_rc = multip_hc_rc.run_hc_rc(uin, dtime=1, height=heights)
+        hidx = land_hc_rc.shape.index(hlvs)
+        for field in land_hc_rc.slices_over(hidx):
```
Comment (gavinevans) [inline lib/improver/tests/wind_downscaling/test_roughnesscorrection.py]: Is there a reason to slice over `hidx` in this case? I would have thought that:
```python
        hidx = land_hc_rc.shape.index(hlvs)
        for field in land_hc_rc.slices_over(hidx):
            self.assertEqual(field.dtype, np.float32)
```
could be replaced by:
```python
self.assertEqual(land_hc_rc.dtype, np.float32)
```

File: lib/improver/wind_downscaling.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def process(self):
                 A 2D array of friction velocities
 
         """
-        ustar = np.full(self.u_href.shape, RMDI)
+        ustar = np.full(self.u_href.shape, RMDI, dtype=np.float32)
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: This seems to be a suitable approach for now. Based on the discussion inhttps://stackoverflow.com/questions/5350342/can-i-set-float128-as-the-standard-float-array-in-numpy and https://stackoverflow.com/questions/5721831/python-making-numpy-default-to-float32, there doesn't seem to be a way to set the default dtype for numpy arrays more globally, so we might need to explore writing our own wrapper functions later.

File: lib/improver/wind_downscaling.py [hunk 3]
```diff
@@ -652,8 +661,14 @@ def __init__(self, a_over_s_cube, sigma_cube, pporo_cube,
         self.sigma = next(sigma_cube.slices([y_name, x_name]))
         try:
             self.z_0 = next(z0_cube.slices([y_name, x_name]))
+            self.z_0.data = np.float32(self.z_0.data)
         except AttributeError:
             self.z_0 = z0_cube
+
+        if height_levels_cube is not None:
+            height_levels_cube.data = height_levels_cube.data.astype(
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: This could be written as, which might be slightly better:
```python
height_levels_cube.data = (
    height_levels_cube.data.astype(np.float32))
```

**Comment by MarkWorsfold** [comment]
Added Gavins changes.
Created some new KGOs and checked that the CLI tests run as expected.
Changed the structure of some equations on pylints suggestions.

File: lib/improver/wind_downscaling.py [hunk 4]
```diff
@@ -914,6 +924,7 @@ def process(self, input_cube):
         if not isinstance(input_cube, iris.cube.Cube):
             msg = "wind input is not a cube, but {}"
             raise TypeError(msg.format(type(input_cube)))
+        input_cube.data = input_cube.data.astype(np.float32)
```
Comment (cgsandford) [inline lib/improver/wind_downscaling.py]: Again, recommend "if cube.dtype == np.float64" before doing this, particularly if the cube has several dimensions.  Also an inline comment that we are doing this for optimisation purposes.

File: lib/improver/tests/wind_downscaling/test_RoughnessCorrection.py [hunk 5]
```diff
@@ -646,6 +648,12 @@ def test_section1f(self):
         land_hc_rc = landpointtests_hc_rc.run_hc_rc(self.uin)
         self.assertArrayEqual(landpointtests_hc_rc.w_cube, land_hc_rc)
 
+    def test_section1g(self):
+        """ Test that code returns float32 precision. """
```
Comment (cgsandford) [inline lib/improver/tests/wind_downscaling/test_RoughnessCorrection.py]: Please remove the space after the first triple quotes - this is to do with how the documentation renders in Sphinx (see https://github.com/metoppv/improver/wiki/Code-Style-Guide#unit-test-structure)

File: lib/improver/tests/wind_downscaling/test_RoughnessCorrection.py [hunk 6]
```diff
@@ -729,6 +737,15 @@ def test_section2c(self):
             _ = multip_hc_rc.run_hc_rc([uin, uin], dtime=2, height=heights,
                                        aslist=True)
 
+    def test_section2d(self):
+        """ Test whether output is float32. """
```
Comment (cgsandford) [inline lib/improver/tests/wind_downscaling/test_RoughnessCorrection.py]: As above.

File: lib/improver/wind_downscaling.py [hunk 7]
```diff
@@ -109,13 +109,10 @@ def process(self):
...
-        ustar = np.full(self.u_href.shape, RMDI)
-        ustar[self.mask] = (
-                VONKARMAN * (
-                    self.u_href[self.mask] /
-                    np.log(self.h_ref[self.mask] / self.z_0[self.mask])
-                )
-        )
+        ustar = np.full(self.u_href.shape, RMDI, dtype=np.float32)
+        numerator = self.u_href[self.mask]
+        denominator = np.log(self.h_ref[self.mask] / self.z_0[self.mask])
```
[the remark this answers is not in the document]
Comment (MarkWorsfold) [reply lib/improver/wind_downscaling.py]: Good point about the log(0) handling. I did a test passing arrays filled with zeros to the equation and it returned an array filled with NaN values - so the code didn't break there.
In the function def _setmask(self): the code creates a mask to mask out zero values - so the situation of passing zeros to this function shouldn't happen anyway.

File: lib/improver/wind_downscaling.py [hunk 8]
```diff
@@ -642,6 +641,11 @@ def __init__(self, a_over_s_cube, sigma_cube, pporo_cube,
             any RC
 
         """
+        a_over_s_cube.data = a_over_s_cube.data.astype(np.float32)
+        sigma_cube.data = sigma_cube.data.astype(np.float32)
+        pporo_cube.data = pporo_cube.data.astype(np.float32)
+        modoro_cube.data = modoro_cube.data.astype(np.float32)
```
[the remark this answers is not in the document]
Comment (MarkWorsfold) [reply lib/improver/wind_downscaling.py]: I've created a new function to do this - Gavin has suggested building a common utility to do this as well. I will talk to him about it.

**Comment by MarkWorsfold** [comment]
I've added Carolines comments - however I am in discussion with Gavin about creating a common utility so that the functionality for float32 downscaling can be used elsewhere.

File: lib/improver/tests/utilities/test_cube_manipulation.py [hunk 9]
```diff
@@ -1709,6 +1710,39 @@ def test_coordinate_raise_exception(self):
...
+        """Create two temperature cubes to test with."""
+        self.cube1 = set_up_temperature_cube()
+        self.cube2 = set_up_temperature_cube()
+
+    def test_basic(self):
+        """Test that the function will return a single iris.cube.Cube with
+           float32 precision."""
+        result1 = self.cube1
+        enforce_float32_precision(result1)
+        self.assertEqual(result1.dtype, np.float32)
```
Comment (gavinevans) [inline lib/improver/tests/utilities/test_cube_manipulation.py]: I think usually `assertIsInstance` is used for testing the type of a variable. This applies to the other unit tests here as well.

File: lib/improver/utilities/cube_manipulation.py [hunk 10]
```diff
@@ -835,6 +835,26 @@ def enforce_coordinate_ordering(
     return cube
 
 
+def enforce_float32_precision(input_cubes):
+    """Take input cube of any precision and convert to float32.
+
+    Args:
+        input_cubes (list of cubes or single cube):
```
Comment (gavinevans) [inline lib/improver/utilities/cube_manipulation.py]: I think this should be as shown below, as list the brackets should contain a python type e.g. list, np.array etc:
```python
input_cubes (list):
```

File: lib/improver/utilities/cube_manipulation.py [hunk 11]
```diff
@@ -835,6 +835,26 @@ def enforce_coordinate_ordering(
...
+            List containing one or more iris cubes to test if not float32
+            precision and downscale to float32 if necessary.
```
Comment (gavinevans) [inline lib/improver/utilities/cube_manipulation.py]: You should mention here something like: `If a list item is not an Iris cube, then this item is skipped.`

File: lib/improver/utilities/cube_manipulation.py [hunk 12]
```diff
@@ -835,6 +835,26 @@ def enforce_coordinate_ordering(
...
+            Note: The code will modify the cubes themselves.
```
Comment (gavinevans) [inline lib/improver/utilities/cube_manipulation.py]: It might be clearest to add `in-place` here.

File: lib/improver/wind_downscaling.py [hunk 13]
```diff
@@ -724,7 +729,7 @@ def calc_av_ppgrid_res(self, a_cube):
                 Cube to calculate average resolution of
 
         Returns:
-            float:
+            float32:
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: I think this should be np.float32.

File: lib/improver/wind_downscaling.py [hunk 14]
```diff
@@ -336,13 +337,16 @@ def calc_roughness_correction(self, hgrid, uold, mask):
         """Function to perform the roughness correction.
 
         Args:
-            hgrid: 3D or 1D np.array (float)
+            hgrid: 3D or 1D np.array (float32)
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: In the IMPROVER documentation, i.e. http://improver.readthedocs.io/en/latest/improver.utilities.cube_manipulation.html#module-improver.utilities.cube_manipulation, it works best if the contents of the brackets is an actual python type. I think this should be `np.float32`, rather than `float32`. Similarly for other uses of `float32` within this program.

File: lib/improver/wind_downscaling.py [hunk 15]
```diff
@@ -336,13 +337,16 @@ def calc_roughness_correction(self, hgrid, uold, mask):
...
                 height above orography
-            uold: 3D np.array (float)
+            uold: 3D np.array (float32)
                 original velocities at hgrid.
+            mask: 2D np.array logical
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: The correct format for docstrings within IMPROVER are documented within the [Code Style Guide](https://github.com/metoppv/improver/wiki/Code-Style-Guide) i.e.
```python
    name_of_variable (variable_type):
```
For example:
```python
    mask (bool):
```
We've adopted this style so that it can be interpreted by Sphinx to create the online documentation.
Could you update this, so that it matches this style? The sentence below could contain a mention that the input array will be 2D e.g.
```python
    """
    mask (bool):
    2D numpy array that is True for land-points, False for Sea and False for
    invalid z_0.
    """
```

File: lib/improver/wind_downscaling.py [hunk 16]
```diff
@@ -641,10 +645,17 @@ def __init__(self, a_over_s_cube, sigma_cube, pporo_cube,
...
-        modoro_cube.data = modoro_cube.data.astype(np.float32)
+        enforce_float32_precision([a_over_s_cube,
+                                   sigma_cube,
+                                   pporo_cube,
+                                   modoro_cube,
+                                   z0_cube,
+                                   height_levels_cube])
+
+        # Standard Python 'float' type is either single or double depending on
+        # system and no reliable method of finding which from variable. So
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: Do you mean: `Standard Python 'float' type is either single or double depending on system and there is no reliable method of finding which from variable`?

File: lib/improver/tests/utilities/test_cube_manipulation.py (hunk 9 above)
Comment (MarkWorsfold) [reply lib/improver/tests/utilities/test_cube_manipulation.py]: We tried this and it didn't accept np.float32 as an argument. So use origonal code

File: lib/improver/tests/utilities/test_cube_manipulation.py [hunk 17]
```diff
@@ -1709,6 +1710,39 @@ def test_coordinate_raise_exception(self):
...
+        self.assertEqual(result2.dtype, np.float32)
+
+    def test_process_none(self):
+        """Test that the function ignores None types."""
+        result1 = self.cube1
+        result2 = None
+        enforce_float32_precision([result1, result2])
+        self.assertEqual(result1.dtype, np.float32)
+        self.assertEqual(result2, None)
```
Comment (cgsandford) [inline lib/improver/tests/utilities/test_cube_manipulation.py]: Better to use `assertIsNone` here.

File: lib/improver/wind_downscaling.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/wind_downscaling.py]: Could you add some explicit unit tests for the `FrictionVelocity` class to cover zero inputs please.  The `_setmask()` function is part of the `RoughnessCorrection` class, and there's no inheritance structure that prevents `FrictionVelocity` from being passed an invalid mask when called from other locations.

**Comment by benfitzpatrick** [comment]
Looks great!

Is it now faster? (E.g. CLI acceptance test speed?)

**Comment by MarkWorsfold** [comment]
@benfitzpatrick The CLI tests you requested are on the JIRA ticket.

File: lib/improver/tests/utilities/test_cube_manipulation.py [hunk 18]
```diff
@@ -1709,6 +1710,39 @@ def test_coordinate_raise_exception(self):
...
+        self.assertEqual(result1.dtype, np.float32)
+        self.assertEqual(result2.dtype, np.float32)
+
+    def test_process_none(self):
+        """Test that the function ignores None types."""
+        result1 = self.cube1
+        result2 = None
+        enforce_float32_precision([result1, result2])
+        self.assertEqual(result1.dtype, np.float32)
+        self.assertIsNone(result2, None)
```
Comment (gavinevans) [inline lib/improver/tests/utilities/test_cube_manipulation.py]: This should be:
```python
self.assertIsNone(result2)
```

File: lib/improver/wind_downscaling.py [hunk 19]
```diff
@@ -203,10 +205,10 @@ def _setmask(self):
 
         Returns:
             (tuple) : tuple containing:
-                **hcmask** (2D array logical):
-                    True for land-points, false for Sea (HC)
-                **rcmask** (2D array logical):
-                    additionally False for invalid z_0 (RC)
+                **hcmask** (bool):
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: I think this should probably be:
```python
    **hcmask** (np.ndarray):
        2D array of booleans etc
```

This applies to other `mask` variables within this file, where the type should be np.ndarray.

File: lib/improver/wind_downscaling.py [hunk 20]
```diff
@@ -445,20 +449,21 @@ def _interpolate_1d(xup, xlow, at_x, yup, ylow):
...
-            xup: 2D np.array (float32)
-                upper x-bins
-            xlow: 2D np.array (float32)
-                lower x-bins
-            at_x: 2D np.array (float32)
+            xup (np.ndarray):
+                2D array float32 - upper x-bins
+            xlow (np.ndarray):
+                2D array float32 - lower x-bins
+            at_x (np.ndarray):
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: It might be worth adding to the sentence below that this is expected to be a 2D array, for consistency with the other docstrings. This applies at `at_x`, `yup` and `ylow`

File: lib/improver/wind_downscaling.py [hunk 21]
```diff
@@ -653,8 +660,8 @@ def __init__(self, a_over_s_cube, sigma_cube, pporo_cube,
                                    height_levels_cube])
 
         # Standard Python 'float' type is either single or double depending on
-        # system and no reliable method of finding which from variable. So
-        # force to numpy.float32 by default.
+        # system and there is no reliable method of finding which from
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: Maybe this should say `from the variable.`

File: lib/improver/wind_downscaling.py [hunk 22]
```diff
@@ -66,13 +67,13 @@ def __init__(self, u_href, h_ref, z_0, mask):
...
-            z_0:  (np.ndarray)
-                A 2D array of floats for the vegetative roughness lengths
-            mask: (np.ndarray)
+            u_href (np.ndarray):
+                A 2D array of float32 for the wind speed at h_ref
+            h_ref (np.ndarray):
+                A 2D array of float32 for the reference heights
+            z_0 (np.ndarray):
+                A 2D array of float32 for the vegetative roughness lengths
+            mask (bool):
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: Could this be changed to be:
```python
    mask (np.ndarray):
```

File: lib/improver/wind_downscaling.py [hunk 23]
```diff
@@ -340,15 +343,18 @@ def calc_roughness_correction(self, hgrid, uold, mask):
...
         Args:
-            hgrid: 3D or 1D np.array (float)
-                height above orography
-            uold: 3D np.array (float)
-                original velocities at hgrid.
+            hgrid (np.ndarray):
+                3D or 1D array float32 - height above orography
+            uold (np.ndarray):
+                3D array float32 - original velocities at hgrid.
+            mask (bool):
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: Could this be changed to be:
```python
    mask (np.ndarray):
```

File: lib/improver/wind_downscaling.py [hunk 24]
```diff
@@ -366,33 +372,37 @@ def calc_roughness_correction(self, hgrid, uold, mask):
...
-            hhere: 2D array (float)
-                height grid to interpolate at
-            mask: 2D array (logical)
+            u_in (np.ndarray):
+                3D array float32 - velocity on h_in layer, last dim is height
+            h_in(np.ndarray):
+                3D or 1D array float32 - height layer array
+            hhere (np.ndarray):
+                2D array float32 - height grid to interpolate at
+            mask (bool):
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: Could this be changed to be:
```python
    mask (np.ndarray):
```

File: lib/improver/wind_downscaling.py [hunk 25]
```diff
@@ -506,19 +518,19 @@ def _calc_height_corr(self, u_a, heightg, mask, onemfrac):
...
-            mask: 3D array(logical)
-                Masks the hc_add result
-            onemfrac: currently, scalar = 1.
-                In principle, it is a function of position and height, e.g.
-                a 3D array (float)
+            u_a (np.ndarray):
+                2D array float32 - outer velocity, e.g. velocity at h_ref_orig
+            heightg (np.ndarray):
+                1D or 3D array float32 - heights above orography
+            mask (bool):
```
Comment (gavinevans) [inline lib/improver/wind_downscaling.py]: Could this be changed to be:
```python
    mask (np.ndarray):
```
You could mention that the contents of the array will be boolean in the docstring.

**Comment by MarkWorsfold** [comment]
@gavinevans  Added your changes to the pull request.
